### PR TITLE
feat: Add article "Creating your custom i18n solution? Don't."

### PR DESCRIPTION
**Pull request by JanCizmar**

### CODE REVIEW

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+description: "Don't create you custom i18n solution. There is a lot of localization tooling. It's reinventing the
+wheel."
+
+tags: [i18n, localization, programming, typescript]
+---
+
+Sometime before, when we did a market analysis, we asked many companies how they solved localization in their project.
+We asked them what's their i18n tooling. Surprisingly a lot of companies
+maintain their custom solution. I've also set up [f5bot](https://f5bot.com/)) to e-mail me when the i18n keyword
```
Comment (stepan662) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:18]: extra ")"

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+But sadly, English is not the only language in the world. There are other languages—for example, Czech - my native
+language.
+There are more plural forms in Czech, so the code would be like this.
+
+```typescript
+if (count === 1) {
+  return 'Peter má jednoho psa.'
+} else if (count >= 2 && count <= 4) {
+  return 'Peter má ${count} psi.'
```
Comment (stepan662) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md]: Petr má ps"y"

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+To get started, you can choose your favorite technology [here](https://tolgee.io/integrations/all) and follow the
+guide.
+
+To be fair, I also have to mention other i18n solutions in the JS world.
+
+### FormatJS (JS, React, Vue)
+
+FormatJs is a set of i18n tools used in JS projects. They maintain
+the [ICU Message Formater port for JS](https://formatjs.io/docs/intl-messageformat). Which we also use in Tolgee to
+format the messages. They have integrations into [React](https://formatjs.io/docs/react-intl)
```
Comment (stepan662) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:215]: integrations "for" React? (I'm not sure)

**Review by stepan662 — APPROVED** [review]

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
+---
+slug: dont-create-your-own-i18n-library
+
+title: "Creating your custom i18n solution? Don't."
+
+authors: [jcizmar]
+
+image: "/img/blog/dont-custom-i18n.png"
+
+description: "Don't create you custom i18n solution. There is a lot of localization tooling. It's reinventing the
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:10]: Meta description should be between 110 and 160 characters

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+wheel."
+
+tags: [i18n, localization, programming, typescript]
+---
+
+Sometime before, when we did a market analysis, we asked many companies how they solved localization in their project.
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:16]: sometime before+adverb 
I would delete sometime before and start the phrase just: When we...

Comment (marketachalupnikova) [reply blog/2022-11-04-dont-create-your-own-i18n-library.md:16]: market analysis in Tolgee

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+When developers think about localizing their projects, some think of i18n only as replacing a string
+with another translated into a selected language. But it's not that trivial. So let me show you what issues you will
+probably face when implementing your custom solution.
+
+## Basic i18n method
+
+As a creator of a custom i18n solution, first of all, you would probably create some method that returns a
+translated string according to provided key. The header of such a function would probably be similar
+to the one below. I am going to use typescript for code examples. However, it would look similar to any other
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:35]: Typescript

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  return data[language]?.[key] || key
+}
+```
+
+OK, cool, that was simple. Now the complicated stuff comes.
+
+## Parameter interpolation
+
+Sometimes we need to provide a parameter for our translation to be rendered. To do so, we have to define some syntax
+to tell which part of the translated sting is a placeholder. Let's make everything wrapped in curly brackets as a
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:65]: string

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+## Plurals
+
+Sometimes you would also like to render strings containing a plural form. For example, consider a
+string `Peter has {count} dogs.`
+This works perfectly when the param `count` is 0, 2, 3, ..., but when the count is 1, the sentence is incorrect.
+
+```
+Peter has 0 dogs.
+Peter has 1 dog. <-- Incorrect!
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:106]: 1 dogs probably?

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+} else {
+  return `Peter has ${count} dog.`
+}
+```
+
+Haha! I don't need any i18n library, you may say. I am the king of my code, you may say!
+
+But sadly, English is not the only language in the world. There are other languages—for example, Czech - my native
+language.
+There are more plural forms in Czech, so the code would be like this.
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:126]: would look

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+The FormatJS port also handles the HTML tags, so you will be safe using that.
+
+## Loading the i18n data
+
+Defining all the data in the `translate` method isn't optimal since the method would rapidly grow in size. Especially in
+JS world, when we are requesting the strings from the browser. We prefer to send only some of the translations in
+all the languages to the client, which probably would never need all the languages. So we must find an effective way
+to provide the language data only when required.
+
+Also, storing the data in JS format is not pretty flexible when letting external translators translate our
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:179]: your app

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+## Recommended way
+
+If this article convinced you not to reinvent the wheel and use the existing i18n library, then you're asking what the
+recommended tools are. The answer is: It depends.
+
+### Tolgee (JS, React, Vue, Angular, Svelte, ...)
+
+For web-based projects, I recommend using Tolgee (surprisingly 😆). It handles everything I mentioned above
+but also provides a few excellent killer features. It's a combination of TMS (localization platform) and i18n library.
+Out of the box, it has an in-context editing feature, which enables you to modify the strings directly in your web app,
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:202]: add link to https://tolgee.io/blog/in-context-production

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+even in production. This can save you a lot of time and stress. It also provides a unique way how you can get perfect
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:203]: perfect translation results

File: blog/2022-11-04-dont-create-your-own-i18n-library.md [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+## TL;DR
+
+- Don't create your custom i18n solution. It's more complicated than it seems
+- You would have to handle parameter interpolation, escaping, plural support, data loading, and more
+- Don't store your i18n data in the database
+- Use ICU Message Format
+- There are many popular i18n solutions
+- Tolgee can save you a lot of time
+- When choosing your i18n solution, check how it interpolates parameters, what formats it supports and
```
Comment (marketachalupnikova) [inline blog/2022-11-04-dont-create-your-own-i18n-library.md:254]: and?

**Pull request merged**